rust/rbac-registration/src/cardano/cip509/validation.rs

```diff
@@ -21,659 +21,348 @@
-                "{function_name}, Invalid payment reference key, 0 is not allowed"
-            ));
-            return None;
-        }
-        // CIP-0509 should only be in conway era
-        if let MultiEraTx::Conway(tx) = txn {
-            // Negative indicates reference to tx output
-            if payment_key < 0 {
-                let index = match decremented_index(payment_key.abs()) {
-                    Ok(value) => value,
-                    Err(e) => {
-                        validation_report.push(format!(
-                            "{function_name}, Failed to get index of payment key: {e}"
-                        ));
-                        return None;
-                    },
-                };
-                let outputs = tx.transaction_body.outputs.clone();
-                let witness = match TxWitness::new(&[txn.clone()]) {
-                    Ok(witnesses) => witnesses,
-                    Err(e) => {
-                        validation_report
-                            .push(format!("{function_name}, Failed to create TxWitness: {e}"));
-                        return None;
-                    },
-                };
-
-                if let Some(output) = outputs.get(index) {
-                    match output {
-                        pallas::ledger::primitives::conway::PseudoTransactionOutput::Legacy(o) => {
-                            return validate_payment_output_key_helper(
-                                &o.address.to_vec(),
-                                validation_report,
-                                &witness,
-                            );
-                        },
-                        pallas::ledger::primitives::conway::PseudoTransactionOutput::PostAlonzo(
-                            o,
-                        ) => {
-                            return validate_payment_output_key_helper(
-                                &o.address.to_vec(),
-                                validation_report,
-                                &witness,
-                            );
-                        },
-                    };
+/// Extracts all stake addresses from both X509 and C509 certificates containing in the
+/// given `Cip509` and converts their hashes to bytes.
+fn extract_stake_addresses(uris: Option<&Cip0134UriSet>) -> Vec<VKeyHash> {
+    let Some(uris) = uris else {
+        return Vec::new();
+    };
+
+    uris.x_uris()
+        .iter()
+        .chain(uris.c_uris())
+        .flat_map(|(_index, uris)| uris.iter())
+        .filter_map(|uri| {
+            if let Address::Stake(a) = uri.address() {
+                a.payload().as_hash().as_slice().try_into().ok()
+            } else {
+                None
+            }
+        })
+        .collect()
+}
+
+/// Checks that only role 0 uses certificates with zero index.
+#[allow(clippy::similar_names)]
+pub fn validate_role_data(metadata: &Cip509RbacMetadata, report: &ProblemReport) {
```

Comment:
For None role 0, 
if there is Some x509 or is Some c509 or is Some pubkey , the first index should be set to undefined

Should not able to do delete in index 0 since it is consider an update to role 0